docroot/sites/registrar.uiowa.edu/modules/registrar_core/src/Plugin/Block/CourseDeadlinesBlock.php

```diff
@@ -0,0 +1,570 @@
+        'lastDayToAddDropNoFee' => 'Last day to add or drop the course without $12 charge',
+        'lastDayToAddWithoutDeanApproval' => "Last day to add without collegiate approval",
+        'lastDayToDropWithoutW' => 'Last day to drop without a "W"',
+        'lastDayToDropWithoutDeanApprovalUndergrad' => "Last day to drop without collegiate approval, undergraduate",
+        'lastDayToDropWithoutDeanApprovalGrad' => "Last day to drop without collegiate approval, graduate",
+      ];
+
+      // Remove $12 charge field for sessions after Fall 2016.
+      // This field is still being populated in MAUI
+      // but is no longer valid.
+      if ($session >= '20163') {
+        unset($dates['lastDayToAddDropNoFee']);
+      }
+
+      foreach ($dates as $key => $label) {
+        if ($data->$key) {
+          $items[] = [
+            '#markup' => $this->t('<span class="uiowa-maui-deadline-label">@label</span> <span class="uiowa-maui-deadline-date">@date</span>', [
+              '@label' => $label,
+              '@date' => date('m/d/Y', strtotime($data->$key)),
+            ]),
+          ];
+        }
+      }
+
+      $deadlines['deadlines_wrapper'] = [
+        '#type' => 'container',
+        '#suffix' => '<p style="text-align: center;"><a href="https://registrar.uiowa.edu/collegiate-office-contact-information-students">Collegiate Office Contact Information</a> to obtain collegiate approval.</p>',
+        '#attributes' => [
+          'class' => ['uiowa-maui-course-deadlines-wrapper', 'deadlines-wrapper'],
+        ],
+        'deadlines' => [
+          '#title' => 'Deadlines',
+          '#items' => $items,
+          '#theme' => 'item_list',
+          '#attributes' => [
+            'class' => 'uiowa-maui-deadlines',
+          ],
+        ],
+      ];
+
+      $times_and_locations = [];
+
+      foreach ($data->timeAndLocations as $key => $value) {
+        if ($value->arrangedTime) {
+          $time = 'Time is to be arranged.';
+        }
+        else {
+          $time = $this->t('@days @start@end', [
+            '@days' => $value->days ?? '',
+            '@start' => $value->startTime ?? '',
+            '@end' => ($value->endTime) ? '-' . $value->endTime : '',
+          ]);
+        }
+
+        $location = NULL;
+
+        if ($value->arrangedLocation) {
+          $location = 'Location is to be arranged.';
+        }
+        elseif ($value->offsite) {
+          $location = $this->t('@building @street @city @country', [
+            '@building' => $value->offsiteBuilding ?? '',
+            '@street' => $value->offsiteStreet ?? '',
+            '@city' => $value->offsiteCity ?? '',
+            '@country' => $value->offsiteCountry ?? '',
+          ]);
+        }
+        else {
+          if ($value->building) {
+            $lookup = json_decode(file_get_contents('https://data.its.uiowa.edu/maps/number-lookup'));
```

Comment:
I think this is okay to launch with but we should revisit this as the list on data.its as it is static and hasn't been maintained.

Reply:
I added a `@todo` in the code for this, and added a quick check so that, if it is out of date and cannot find a building code, it will still output the location sans link.